### ISSUE TITLE
Proposal: Introduce Shell as the default way of using/learning .NET MAUI (templates, documentation, etc.)

### ISSUE THREAD

**Issue by jamesmontemagno** [dotnet/maui#4522]
### Description

The Shell concept was introduced in Xamarin.Forms 4.0 and has been brought over to .NET MAUI and in my opinion helps standardize application structure, navigation, theming, and at the same time abstracts away complicated structure of Flyout and Tabs. Today, developers coming to Xamarin.Forms and .NET MAUI for the first time will have to make a decision if they should use Shell or not, which has rippling effects down the road. The default templates for Xamarin.Forms have mostly been updated to default to Shell, however Blank app remains a non-Shell based application, and the documentation separates out Shell into a different section, which means that you have to know to go look for it to go learn the concepts. I believe that developers will be more successful learning and using .NET MAUI if Shell is the default way of building applications, learning applications, and is the default template when running `dotnet new maui`.

Benefits:
* Standard way of learning app structure, navigation, theming, and more
* URI navigation & constructor injection simplify development
* Integrated search handling
* Documentation can be simplified by focusing on Shell concepts first
* This is our chance to do it right and not make things confusing with multiple choices, make Shell the way.

I am not asking Shell to be required in .NET MAUI, so non-Shell based XF apps can easily migrated over and traditional navigation can still work just fine. But, I believe that Shell was created to make cross-platform development easier and from my personal usage it does.

Here is my PR for proposal changes to templates -> https://github.com/dotnet/maui/pull/4524

### Public API Changes

None, we would update the templates to include a Shell with theming and add a single page.

### Intended Use-Case

Make app development with .NET MAUI easier than ever, while still be backwards compatible.

**Comment by PrestigeDevop** [comment]
Yes please!  , also dynamic shell .

**Comment by CliffAgius** [comment]
Can't argue with that, makes onboarding easier for those coming from other .NET backgrounds whats not to like.

Maybe have a template of Empty/Bare MAUI for those that don't like or can't use shell for some reason and you have all the bases covered.

Make it happen @jamesmontemagno 😜

**Comment by Chase-William** [comment]
I concur that unification of Shell into MAUI as the standard approach for the various aspects mentioned by James would be beneficial; especially for students.

> A few years back I was a student learning mobile app development through Xamarin.Forms. I was building apps for school and personal exploration and often found myself annoyed by the boiler plate code/files required to get a `drawer navigation` going. I discovered Shell for Xamarin.Forms and tried using it, at the time it was still in a *experimental* state. Using Shell helped reduce the boiler plate code in my app and its use was intuitive, but I ultimately stopped because it wasn't *ready* yet. My point is, as a student it was improving my development experience, but at the same, it's use in my apps concerned me because of it's separation from the norm in XF *(at that time)*.

Therefore, from my experience, I think @jamesmontemagno's proposal would be beneficial to students learning mobile app development with MAUI.

**Comment by scearces** [comment]
Good idea!

**Comment by jamesmontemagno** [comment]
> Maybe have a template of Empty/Bare MAUI for those that don't like or can't use shell for some reason and you have all the bases covered.

Absolutely! I think just 1 template is the way to go, I don't think there should be any more complexity than that as too many choices are bad.

Some feedback from Twitter (https://twitter.com/JamesMontemagno/status/1490027025289076740) mostly has been around giving the ability to navigate with Shell by also passing complex data types instead of urls. You can do this already today via the built in INavigation, but I think we should also just add this to Shell as an overload.

**Comment by SmartmanApps** [comment]
As a teacher, please keep default to the minimal app that will work. Sure, provide shell as a radio-button choice, but not default in a way like the current "blank" app has XAML in it which then needs to then be stripped out, and there's no XAML-free choice provided. People who are learning need to be taught 1 step at a time - start with the basics, then add further things on top. I have taught high school students Xamarin/C#, and so as to keep it simple for them, the first thing I actually had to do was teach them how to strip out the XAML, so that we could just focus on Xamarin/C# until they got a handle on it. Once they understand that THEN you can teach them XAML as well. The golden rule is one step at a time (or as few steps at a time as possible). To illustrate the issue with learners, when asked to install Visual Studio, one of them installed Blend for Visual Studio. Another time I said they could rename App1 to whatever they wanted to call their app, so one of them changed it to just App - obviously that created a whole bunch of problems until I worked out what she'd done. K.I.S.S.

If we're adding templates then can we please have...
- a C# blank app (no XAML)
- a C# Flyout with no extra features thrown in for demo purposes - JUST the skeleton Flyout page, just C#.

The first I always have to strip out all the XAML first. The second I don't even use - so much needs to be stripped out it's easier to just start from a blank app and add what you need (and even that does take a while. I've ended up building a base-class for it to save me some time).

**Comment by jamesmontemagno** [comment]
@SmartmanApps I don't want to add more templates. I want Shell to be the way people learn and is THE WAY to build apps. So there isn't more complexity at all.  It should set people up for success and be extremely minimal. 

There is 1 template, which adds shell as the structure for the app with Shell and has the same MainPage we know and love.

https://github.com/dotnet/maui/pull/4524 is the PR and you can see you could just delete AppShell.xaml and AppShell.xaml.cs and make the MainPage the startup instead of AppShell.

**Comment by dotMorten** [comment]
I think the learning argument is a bit misguided. IMHO templates shouldn’t be used for learning but for getting a project started. Learning is something you do once. Starting a project is something you do a lot.

I’d also worry about ending up with all apps looking the same if you premake too much. Should there be a shell template? Definitely! Should it be default? IMHO no.

And wrt learning I think it would be easier for the first app to be as simple as possible and build up to using shell.

**Comment by jonlipsky** [comment]
I agree with @dotMorten.  When I create a blank app, I want a truly  blank app.

**Comment by davidbritch** [comment]
It seems to me that rather than make a change such as this based on opinion, we should have data from A|B testing new users (rather than twitter polls/surveys of existing users) that indicates:

- A Shell-first approach improves user acquisition/retention and/or learning outcomes for new users.
- Shell is the right approach for desktop users. Do desktop users think that a tab/flyout approach is the correct approach to navigation in their apps, rather than a menu based approach. Ultimately, is a Shell-first approach correct for desktop users?

**Comment by fazary** [comment]
does it still lagging when we press flyout item like xamarin forms ?

**Comment by Jared-Aus** [comment]
So, after ~15 years of XAML and view models this is where navigation is at?
`await Shell.Current.GoToAsync($"{nameof(ItemDetailPage)}?{nameof(ItemDetailViewModel.ItemId)}={item.Id}");`

First wondering how to inject this Shell.Current dependency thingy into the view model and what’s going to happen in a unit test if I don’t, followed by thoughts of why is a view model referencing a page?

Then, in the ItemDetailViewModel, instead of requiring _item_ as the main and perhaps only dependency of the class in the constructor, one needs to replace it with property injection using special attributes and a public property that’s not likely to be even used indirectly by the view.

In the meantime, the page and view model have been constructed, the binding context set, and all the XAML bindings have run against your view model - which at this stage still knows nothing about the item it’s meant to display.

Eventually the view model will receive the id so you can fetch the object you already had when you started the navigation. But first, you’ll need to add to the constructor the data fetching service you likely didn’t need before.

Now you’ve finally retrieved a reference to the item, all the view bound properties will need to be either set or manually notified of changes so the view can be practically redrawn a second time. Most likely bad luck to you if you thought you’d be clever and save overheads with a sprinkling of one-time bindings etc….

Maybe this all makes complete sense in a ASP.Net web request/response world and also roughly with Blazor, but from what I’m seeing at present (and based on XF experience with it) Shell navigation is likely too complex for beginners and too inflexible for advanced use. I really hope I’m wrong though.

**Comment by Beelzenef** [comment]
I can only say... yes please! 😍

**Comment by pictos** [comment]
> have been looking at Shell in Forms and MAUI but still not sure it suits 'line of business apps' for companies. I develop apps with specific functions that should be completed before moving to another task. The Shell design with flyouts works against this and allows users to flip between tasks leaving tasks forgotten and incomplete. I like the Shell navigation as it works like Prism which we have used in the past. Once MAUI is out of preview I hope to revisit Shell and try to create a more rigid UX for business apps.

@lywyn, Shell already provides a way to cancel the navigation if something is incomplete on the flow, and supports async scenarios. Take another look at it, it gets better (;

@Jared-Aus 
> First wondering how to inject this Shell.Current dependency thingy into the view model and what’s going to happen in a unit test if I don’t, followed by thoughts of why is a view model referencing a page?

Just like you do with the normal `PushAsync`, `PopAsync`, etc... You need an instance of a NavigationPage to do it, but you can abstract that using a service. Just do it for Shell as well.

> Then, in the ItemDetailViewModel, instead of requiring item as the main and perhaps only dependency of the class in the constructor, one needs to replace it with property injection using special attributes and a public property that’s not likely to be even used indirectly by the view.

You can use an [interface](https://docs.microsoft.com/en-us/xamarin/xamarin-forms/app-fundamentals/shell/navigation#process-navigation-data-using-a-single-method) instead of an attribute.

> In the meantime, the page and view model have been constructed, the binding context set, and all the XAML bindings have run against your view model - which at this stage still knows nothing about the item it’s meant to display.

> Eventually the view model will receive the id so you can fetch the object you already had when you started the navigation. But first, you’ll need to add to the constructor the data fetching service you likely didn’t need before.

> Now you’ve finally retrieved a reference to the item, all the view bound properties will need to be either set or manually notified of changes so the view can be practically redrawn a second time. Most likely bad luck to you if you thought you’d be clever and save overheads with a sprinkling of one-time bindings etc….

If I understood correctly, this is flow is the same if you use an MVVM framework that supports navigation data, like Prism and his `OnNavigatedTo`/`OnNavigatedFrom` overrides, we have had that for a long time and never heard any dev complaining about it.

**Comment by charlesroddie** [comment]
It is a mistake to keep Shell integrated in MAUI rather than as an add-on, and this is compounded by making Shell the default. Failure to learn the lessons of Xamarin.Forms.

1. Shell is an ultra-low performance framework currently: adds 0.5s to android startup https://github.com/dotnet/maui/pull/5111 . So as soon as your app gets close to production you will need to switch out of Shell.
2. Shell bugs and MAUI bugs will be hard to distinguish. You will have reproductions of core bugs using shell-based code: users will think that since there is a default template with Shell, a shell-based reproduction will count as a minimal reproduction. This will make problems very difficult to isolate and gives little confidence that the Xamarin.Forms situation of unmanageable issues will by solved in MAUI.
3. Navigation is not the core of MAUI. Giving MAUI more responsibilities will end up recreating Xamarin.Forms - having a buggy framework which tries to do everything. Focusing on the one issue that needs solving, namely taking structured objects in dotnet and laying them out in platforms.
